### PR TITLE
[QUIC] Added tests to check IsSupported.

### PR DESCRIPTION
**Pull request by ManickaP**
On Windows, straightforward.

On Linux, I'm probing system for libmsquic existence and if so then checking `IsSupported`. This should discover installed libmsquic without corresponding dependencies like openssl.
I'm also adding test that expects libmsquic on every tested Linux we have in the matrix and suppressing for known gaps.

Contributes to #81447

### CODE REVIEW

**Comment by ManickaP** [comment]
Thinking about this some more and I think we should be more aggressive. On Linux, on every docker image, we should expect `libmsquic`. If it's not there, we should add it. If we cannot, we should file an issue and disable the test temporarily, e.g. https://github.com/dotnet/dotnet-docker/issues/4322#issuecomment-1400840703.

File: src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicPlatformDetectionTests.cs [hunk 1]
```diff
@@ -19,5 +20,33 @@ public void UnsupportedPlatforms_ThrowsPlatformNotSupportedException()
...
+            string output = await ldconfig.StandardOutput.ReadToEndAsync();
+            await ldconfig.WaitForExitAsync();
+            if (output.Contains("libmsquic.so"))
+            {
+                Assert.True(QuicListener.IsSupported);
+                Assert.True(QuicConnection.IsSupported);
+            }
+            else
+            {
+                _output.WriteLine("No msquic library found.");
```
Comment (carlossanlop) [inline src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicPlatformDetectionTests.cs:48]: If the above `if` condition emits two assert trues, shouldn't this `else` output an assert false?

Comment (carlossanlop) [reply src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicPlatformDetectionTests.cs:48]: Nevermind. I was reading it wrong. I see what you meant.

File: src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicPlatformDetectionTests.cs [hunk 2]
```diff
@@ -19,5 +20,33 @@ public void UnsupportedPlatforms_ThrowsPlatformNotSupportedException()
...
+            Assert.True(QuicListener.IsSupported);
+            Assert.True(QuicConnection.IsSupported);
+        }
+
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsLinux))]
+        public async Task SupportedLinuxPlatforms_IsSupportedIsTrue()
+        {
+            using Process ldconfig = new Process();
+            ldconfig.StartInfo.FileName = "ldconfig";
+            ldconfig.StartInfo.Arguments = "-p";
```
Comment (wfurt) [inline src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicPlatformDetectionTests.cs:36]: This is going to fail on Alpine

docker run -ti mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.14-helix-amd64 bash
```
bash-5.1$ ldconfig -p
Illegal option -p
```

Comment (wfurt) [reply src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicPlatformDetectionTests.cs:36]: I'm not sure what would be better option. Perhaps pinvoke `dlopen()` ?

Comment (ManickaP) [reply src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicPlatformDetectionTests.cs]: I switched that to `find` and it seems like it caught first problem with alpine:
```
 Discovering: System.Net.Quic.Functional.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Net.Quic.Functional.Tests (found 3 of 114 test cases)
  Starting:    System.Net.Quic.Functional.Tests (parallel test collections = on, max threads = 2)
    System.Net.Quic.Tests.MsQuicPlatformDetectionTests.SupportedWindowsPlatforms_IsSupportedIsTrue [SKIP]
      Condition(s) not met: "IsWindows"
    System.Net.Quic.Tests.MsQuicPlatformDetectionTests.SupportedLinuxPlatforms_IsSupportedIsTrue [FAIL]
      Assert.True() Failure
      Expected: True
      Actual:   False
      Stack Trace:
        /_/src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicPlatformDetectionTests.cs(45,0): at System.Net.Quic.Tests.MsQuicPlatformDetectionTests.SupportedLinuxPlatforms_IsSupportedIsTrue()
        --- End of stack trace from previous location ---
      Output:
        Using unknown
        /usr/lib/libmsquic.so.2
        
  Finished:    System.Net.Quic.Functional.Tests
=== TEST EXECUTION SUMMARY ===
   System.Net.Quic.Functional.Tests  Total: 3, Errors: 0, Failed: 1, Skipped: 1, Time: 0.715s
```
https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-81481-merge-f74ebb134a30429ca3/System.Net.Quic.Functional.Tests/2/console.b17bfdef.log?helixlogtype=result

Comment (ManickaP) [reply src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicPlatformDetectionTests.cs]: Based on my local test, it fails here:
https://github.com/dotnet/runtime/blob/490b20fcfa6c984d04b8163cb81c9b65029f55c5/src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicApi.cs#L69-L70

Comment (ManickaP) [reply src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicPlatformDetectionTests.cs]: Lib is present, but cannot be loaded for some reason. `ldd` fails with:
```
ldd /usr/lib/libmsquic.so.2
        /lib/ld-musl-x86_64.so.1 (0x7fbc0efdf000)
        libcrypto.so.1.1 => /lib/libcrypto.so.1.1 (0x7fbc0ec00000)
        libc.musl-x86_64.so.1 => /lib/ld-musl-x86_64.so.1 (0x7fbc0efdf000)
Error relocating /usr/lib/libmsquic.so.2: EVP_seed_cbc: symbol not found
```

@wfurt could you help here?

Comment (wfurt) [reply src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicPlatformDetectionTests.cs]: yes. good catch. 

https://git.alpinelinux.org/aports/tree/main/openssl/APKBUILD?h=3.15-stable#n110

it seems like the crypto on alpine is built with different options than Quic counter part. I can fix our test images. 

I would assume quic does not care so we can perhaps also update msquic defaults.

File: src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicPlatformDetectionTests.cs [hunk 3]
```diff
@@ -20,20 +21,33 @@ public void UnsupportedPlatforms_ThrowsPlatformNotSupportedException()
             Assert.ThrowsAsync<PlatformNotSupportedException>(async () => await CreateQuicConnection(new IPEndPoint(IPAddress.Loopback, 0)));
         }
 
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/73290", typeof(PlatformDetection), nameof(PlatformDetection.IsSingleFile))]
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicPlatformDetectionTests.cs:24]: Based on this run: https://dev.azure.com/dnceng-public/public/_build/results?buildId=158551&view=ms.vss-test-web.build-test-results-tab&runId=3276022&resultId=132009&paneView=debug
it looks like this part works.

**Comment by ManickaP** [comment]
ATM depends on #81742

**Comment by wfurt** [comment]
> ATM depends on #81742

why?

**Comment by ManickaP** [comment]
> > ATM depends on #81742
> 
> why?

Ohh we don't, you have too many PRs named "update msquic" :smile: 

We need to pull in https://github.com/microsoft/msquic/pull/3394 for alpine, which was merged into transport in https://github.com/dotnet/msquic/pull/120.

Do we know if got propagated into the helix images already?

**Comment by wfurt** [comment]
`alpine-3.17-helix-amd64-20230208010446-6d7b22b` and `alpine-3.15-helix-amd64-20230208010517-6d7b22b` should be ready. However, they did not roll into production yet so it is not possible to use the simplified stamp-less version like `alpine-3.17-helix-amd64`. Since there probably will be more changes we may not wait. 
We can do it as part of this or I was planning to roll it in as contribution to https://github.com/dotnet/runtime/issues/81801.

**Comment by wfurt** [comment]
the image tags are waiting for https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/790. Should hopefully happen today.

File: src/libraries/Common/tests/TestUtilities/System/PlatformDetection.Unix.cs [hunk 4]
```diff
@@ -52,6 +55,7 @@ public static partial class PlatformDetection
         public static bool IsRedHatFamily => IsRedHatFamilyAndVersion();
         public static bool IsNotRedHatFamily => !IsRedHatFamily;
         public static bool IsRedHatFamily7 => IsRedHatFamilyAndVersion(7);
+        public static bool IsCentos7 => IsDistroAndVersion("centos", 7);
```
Comment (wfurt) [inline src/libraries/Common/tests/TestUtilities/System/PlatformDetection.Unix.cs:58]: can we use the `IsRedHatFamily7` above?

Comment (ManickaP) [reply src/libraries/Common/tests/TestUtilities/System/PlatformDetection.Unix.cs:58]: It includes RHEL (or rather excludes from testing) as well. So no, I don't think that's advisable as we want exact list of images that are failing.

**Comment by ManickaP** [comment]
This is ready for review: @CarnaViire @wfurt

**Review by CarnaViire — APPROVED** [review]
LGTM, thanks! :shipit:

File: src/libraries/Common/tests/TestUtilities/System/PlatformDetection.Unix.cs (hunk 4 above)
Comment (wfurt) [reply src/libraries/Common/tests/TestUtilities/System/PlatformDetection.Unix.cs:58]: they should be basically identical. but I guess they are not:
https://github.com/dotnet/arcade/issues/10253#issuecomment-1215320634

Perhaps something we should investigate. We have at least one of the runs in container so it should be easy to check.

**Pull request merged**